README.md

```diff
@@ -0,0 +1,237 @@
+  * Container Core, a library focused on bridging the gaps between Docker and Java
+  * Testcontainers 2.x, a library focused on using Container Core for testing
+* **(Secondary, but important)** Even though there will be breaking API changes for Testcontainers 2.0, make best efforts to ensure a 1:1 mapping of concepts
+
+
+---
+
+##Runtime `Container` vs define-time `ContainerBuilder`
+
+Separating out the API for 'actions that are available when a container is being defined' and 'actions that are available when the container' is running is important. It will help reduce the API surface area to only the methods that make sense in a given context. This should improve ease of use and discoverability of the API, as well as resulting in a cleaner codebase.
+
+The `org.testcontainers.containercore` package shows the key classes for creating containers *that are not tied to any test framework*. Classes under this package would be **published as a ContainerCore JAR** (probably still under the org.testcontainers groupId, but a new artifact)
+
+### End result / usage
+
+The classes in this package allow for the following API and usage:
+
+```java
+// Building a simple container object not linked to any test framework
+final Container container = Container.builder()
+  .withImage("alpine:3.5")
+  .withExposedPort(1234)
+  .build();
+
+// now that the container is built, the `with___` methods disappear and runtime methods,
+//  like `exec` become available
+container.start();
+container.exec("date");
+container.stop();
+```
+
+
+
+### Implementation
+
+* `Container` holds configuration data in fields, plus methods for 'run-time' interactions such as `start`, `stop` , `exec`, etc.
+* `ContainerBuilder` (and its parent class `BaseContainerBuilder`)  exposes the define-time `withXYZ` fluent setter API
+* `BaseContainerBuilder` uses package-scoped access to `Container`'s configuration fields to apply values before `build()` is called
+* `ContainerBuilder` can also validate the container's configuration just prior to building, e.g. to ensure that essential fields like `image` are set
+
+
+
+
+---
+
+## Test-framework agnostic vs Test-framework coupled classes
+
+The above example shows test-framework agnostic usage. Still to support use of JUnit 4 rules, we need an alternative Container implementation (implementing JUnit's `TestRule`) and an alternative Builder specialisation. 
+
+### End result / usage (JUnit 4 case)
+
+See the package `org.testcontainers.testing.junit4` - **this would be published as a Testcontainers JAR**.
+
+The end result achieved is almost the same as the existing Testcontainers 1.x API. The only differences are usage of a different class name and use of the builder pattern.
+
+```java
+public class SimpleJUnit4Test {
+
+    @Rule
+    public ContainerTestRule alpine = ContainerTestRule.builder()
```

Comment:
If we do it this way, we will have to duplicate each and every `withXXX` method. What if we use something like I proposed with Networks? i.e.:
```java
Container.builder()
   .withSomething()
   .build(asJUnit4Rule());

Container.builder()
   .withSomething()
   .build(asJUnit5Rule());

Container.builder()
   .withSomething()
   .build(); // aka GenericContainer
```

Reply:
What about composing it the other way around?
Something like:

```
ContainerTestRule.builder(
   Container.builder()
   .withSomething()
).build();
```

Reply:
@kiview I thought about it as well, but (compared to my example) it adds more code around it without providing any visible benefits IMO. Also, if `.build()` accepts an interface, it will be possible to easily discover all available adapters like JUnit4/5 and others

Reply:
+1 for the discoverability, makes sense.

Reply:
Actually as-is we don't have to duplicate the `with` methods... If you look, the [`junit.ContainerTestRuleBuilder`](https://github.com/rnorth/containercore/pull/1/files#diff-7694e5cd7c840b34f8e5e9bae3940c45) class extends `BaseContainerBuilder`, which does the generic self-typed-return thing we currently do in Testcontainers 1.x. (This helps allow for a fluent interface).

[`containercore.ContainerBuilder`](https://github.com/rnorth/containercore/pull/1/files#diff-b6760232d3c1738850b0c0e65fc9ac56) also uses this same base class as well.

As such, we don't need to duplicate the methods. 
The downside is that self-typed-returns are a bit confusing, but at least they're isolated to the builder classes now. I'm not sure how you feel about keeping this technique in use, though.

@bsideup I'm all for the approach you suggest if we don't go for this model above. The only reason I didn't go for that straight away is a slight discoverability worry.

Reply:
TBH I myself don't really understand the <SELF> type hack 😅 
It's for keeping the this type when calling methods from the parent class?

Reply:
Hehe, yes, it's so `SomeSubclass::withSomeMethodFromParent` returns as `SomeSubclass` rather than the parent class. 
I'll consider this as a vote against 😆